vyper/venom/passes/memmerging.py

```diff
@@ -0,0 +1,252 @@
+        index = bisect_left(intervals, inter)
+
+        if index > 0:
+            if intervals[index - 1].overlap(inter):
+                return True
+
+        if index < len(intervals):
+            if intervals[index].overlap(inter):
+                return True
+
+        return False
+
+    def _handle_bb(
+        self, bb: IRBasicBlock, load_inst: str, copy_inst: str, allow_dst_overlaps_src: bool = False
+    ):
+        loads: dict[IRVariable, int] = dict()
+        intervals: list[_Interval] = []
+
+        def _barrier():
+            self._optimize_copy(bb, intervals, copy_inst)
+            loads.clear()
+
+        for inst in bb.instructions.copy():
+            if inst.opcode == load_inst:
+                src_op = inst.operands[0]
+                if not isinstance(src_op, IRLiteral):
+                    _barrier()
+                    continue
+                assert inst.output is not None  # help mypy
+                uses = self.dfg.get_uses(inst.output)
+                if len(uses) != 1:
+                    _barrier()
+                    continue
+                if uses.first().opcode != "mstore":
+                    _barrier()
+                    continue
+                if self._overlap_exist(
+                    intervals, _Interval(src_op.value + 32, src_op.value, 32, [])
+                ):
+                    _barrier()
+                    continue
+                assert isinstance(inst.output, IRVariable)
+                loads[inst.output] = src_op.value
+            elif inst.opcode == "mstore":
+                var = inst.operands[0]
+                dst = inst.operands[1]
+                if not isinstance(dst, IRLiteral) or not isinstance(var, IRVariable):
+                    _barrier()
+                    continue
+                if var not in loads:
+                    _barrier()
+                    continue
+                src: int = loads[var]
+                mload_inst = self.dfg.get_producing_instruction(var)
+                assert mload_inst is not None  # help mypy
+                n_inter = _Interval(dst.value, src, 32, [mload_inst, inst])
+                if not self._add_interval(
+                    intervals, n_inter, allow_dst_overlap_src=allow_dst_overlaps_src
+                ):
+                    _barrier()
+            elif _volatile_memory(inst):
+                _barrier()
+
+        self._optimize_copy(bb, intervals, copy_inst)
+
+    # optimize memzeroing operations
+    def _optimize_memzero(self, bb: IRBasicBlock, intervals: list[_Interval]):
+        for interval in intervals:
+            inst = interval.insts[-1]
+            if interval.length == 32 and inst.opcode == "calldatacopy":
+                dst = inst.operands[2]
+                inst.opcode = "mstore"
+                inst.operands = [IRLiteral(0), dst]
+            elif interval.length <= 32:
+                continue
```

Comment:
i'm not sure we have any examples of this, but we could have multiple small calldatacopies fused together